<img src="https://lh3.googleusercontent.com/drive-viewer/AJc5JmT3-KppjOXJ81Wj6MqcvqmKe2VbLzFP2fJRaXQSvPXxO4JfPi9jrca4-m6yfYWEYZm38rE0M30=w3584-h1706" ><br>





# PROYECTO FINAL - SEGUNDA ENTREGA - MODELO 1

Integrantes:

Juan Sebastian Alvarez Eraso

Rafael Camilo Tejon Rojas

Oscar Javier Angel Balcázar

**Objetivos**

- Finalizar las actividades de entendimiento de los datos y enfoque analítico así como el alcance general del proyecto.
- Realizar la preparación de datos requerida para la construcción de un modelo basado en machine learning.
- Entrenar un primer modelo de machine learning a partir del enfoque analítico
definido y realizar una primera evaluación de resultados.

**Segundo Sprint del proyecto**

● El proyecto se debe enfocar en la preparación de datos y construcción
de un primer modelo predictivo o explicativo basado en regresión, clasificación o alguna otra técnica analítica previamente discutida con los docentes. 



#### **Problemática seleccionada**

Después de una investigación en diferentes temáticas seleccionamos un Dataset con datos de hurtos de la ciudad de Medellín.

**Descripción del Dataset**

Para este taller se utilizarán el siguiente Dataset:


1. Hurto registrados por la Policía Nacional en la ciudad de Medellín<sup>1</sup> cometidos contra personas en espacio público.

 **Diccionario de Datos:**


| Nombre | Tipo |Descripción |
|--------|------|------------|
| SEXO | string  | sexo de la víctima, o cuando lo que se mide es la comisión de un delito del presunto indiciado.|
| EDAD | number |	edad de la víctima, o cuando lo que se mide es la comisión de un delito, será del presunto indiciado.|
| ESTADO_CIVIL | string  | estado civil de la víctima, o cuando lo que se mide es la comisión de un delito, será del presunto indiciado.|
| GRUPO_ACTOR | string  | de la víctima, o cuando lo que se mide es la comisión de un delito, será del presunto indiciado. El grupo actor indica el grupo ej. Guerrilla, Paramilitares etc.|
| ACTIVIDAD_DELICTIVA | string  | actividad ilegal a la que se dedica la víctima o el presunto indiciado. Mirar en contexto con conducta.|
| PARENTESCO | string  | parentesco entre la víctima y el presunto indiciado.|
| OCUPACION | string  | ocupación de la víctima o del presunto indiciado, Mirar en contexto con conducta.|
| DISCAPACIDAD | string  | condición de movilidad reducida de la víctima o indiciado.|
| GRUPO_ESPECIAL | string  | indica grupos especiales, que están asociados de por vida a la persona y que hacen que sean más vulnerables a un hecho en particular.|
| MEDIO_TRANSPORTE | string  | medio transporte donde se movilizaba la víctima o el presunto indiciado según el caso.|
| NIVEL_ACADEMICO | string  | nivel acádemico de la víctima o el presunto indiciado según el caso.|
| TESTIGO | string  | indica si hubo o no un testigo de los hechos.|
| NOMBRE_BARRIO | string  | nombre del barrio donde ocurrieron los hechos.|
| CODIGO_BARRIO | string  | código del barrio donde ocurrieron los hechos.|
| CODIGO_COMUNA | string  | código de la comuna donde ocurrieron los hechos.|
| LUGAR | string  | lugar donde ocurrieron los hechos. El lugar es una tipificación del urbanismo más cercano al hecho.|
| SEDE_RECEPTORA | string  | entidad que recepcionó el caso.|
| CONDUCTA | string  | es el inventario de indicadores o hechos asociados a la seguridad y convivencia.|
| MODALIDAD | string  | es la forma como se materializa el hecho.|
| CARACTERIZACION | string  | variable que permite analizar y tipificar la conducta.|
| CONDUCTA_ESPECIAL | string  | una forma de caracterización adicional.|
| ARMA_MEDIO | string  | es el arma, medio o mecanísmo con el que se comete el hecho.|
| ARTICULO_PENAL | string  | artículos del código penal o del código de contravenciones de Policía.|
| CATEGORIA_PENAL | string  | caracterización de algunos tipos penales.|
| BIEN | string  | nombre del bien.|
| CATEGORIA_BIEN | string  | agrupación de bienes, nivel intermedio de agregación.|
| GRUPO_BIEN | string  | agrupación de bienes, nivel alto de agregación.|
| MODELO | string  | modelo del bien.|
| COLOR | string  | color del bien.|
| PERMISO | string  | tipo de permiso de porte que tiene el bien.|
| UNIDAD_MEDIDA | string  | unidad del bien.|
| FECHA_HECHO	| date | fecha en que ocurrió el hecho.|
| LATITUD	| number |	latitud geográfica sistema de coordenadas wgs84	|
| LONGITUD | number |	longitud geográfica sistema de coordenadas wgs84	|
| CASO | number |	1: indicará el caso, 0: podría indicar bienes o persona, dependiendo de la conducta que se busca analizar, es decir si el hecho es contra la vida y la integridad personal o es contra el patrimonio.|
| VALOR | number |	valor de uso del bien.|
| CANTIDAD | number |	cantidad que debe ser entendida en el contexto de la unidad de medida.|

#### **Lectura y Análisis del Dataset.**

##### Importar librerías

In [ ]:
import pandas as pd
from pandas_profiling import ProfileReport
import pandas as pd
import plotly.express as px
from plotly.offline import init_notebook_mode, iplot, plot
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns

from sklearn.preprocessing import OrdinalEncoder
from sklearn.linear_model import LinearRegression, SGDRegressor
from sklearn.compose import make_column_transformer
from sklearn.preprocessing import StandardScaler, PolynomialFeatures
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, mean_absolute_error
# Configuración de pandas para extender el número de filas y columnas en la visualización
pd.set_option('display.max_rows', 100)
pd.set_option('display.max_columns', 100)

##### Lectura del Dataset

In [ ]:
# Loading the CSV file as dataframe
med_hurto_df = pd.read_csv('https://raw.githubusercontent.com/juanalvarez123/MINE-4101-proyecto-final/main/data/hurto_a_persona_medellin.csv', delimiter=";", error_bad_lines=False)

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:3326: FutureWarning:

The error_bad_lines argument has been deprecated and will be removed in a future version.



/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:3326: DtypeWarning:

Columns (20) have mixed types.Specify dtype option on import or set low_memory=False.



In [ ]:
med_hurto_df.shape

(280870, 31)


***Tipos de Variables***


| Nombre | Tipo de Variable|
|--------|---------|
| SEXO | Cualitativa Nominal|
| EDAD | Cuantitativa Continua  |
| ESTADO_CIVIL | Cualitativa Nominal |
| GRUPO_ACTOR | Cualitativa Nominal|
| ACTIVIDAD_DELICTIVA | Cualitativa Nominal|
| PARENTESCO | Cualitativa Nominal|
| OCUPACION | Cualitativa Nominal|
| DISCAPACIDAD | Cualitativa Nominal|
| GRUPO_ESPECIAL | Cualitativa Nominal|
| MEDIO_TRANSPORTE | Cualitativa Nominal|
| NIVEL_ACADEMICO | Cualitativa Ordinal|
| TESTIGO | Cualitativa Nominal|
| NOMBRE_BARRIO | Cualitativa Nominal|
| CODIGO_BARRIO | Cualitativa Nominal|
| CODIGO_COMUNA | Cualitativa Ordinal|
| LUGAR | Cualitativa Nominal|
| SEDE_RECEPTORA | Cualitativa Nominal|
| CONDUCTA | Cualitativa Nominal|
| MODALIDAD | Cualitativa Nominal|
| CARACTERIZACION | Cualitativa Nominal|
| CONDUCTA_ESPECIAL | Cualitativa Nominal|
| ARMA_MEDIO | Cualitativa Nominal|
| ARTICULO_PENAL | Cualitativa Nominal|
| CATEGORIA_PENAL | Cualitativa Nominal|
| BIEN | Cualitativa Nominal|
| CATEGORIA_BIEN | Cualitativa Nominal|
| GRUPO_BIEN | Cualitativa Nominal|
| MODELO | Cualitativa Nominal|
| COLOR | Cualitativa Nominal|
| PERMISO | Cualitativa Nominal|
| UNIDAD_MEDIDA | Cualitativa Nominal|
| FECHA_HECHO	| Cualitativa Ordinal|
| LATITUD	| Cuantitativa Continua|
| LONGITUD	| Cuantitativa Continua|
| CASO | Cualitativa Nominal|
| VALOR | Cualitativa Ordinal|
| CANTIDAD | Cualitativa Ordinal|

# ***Limpieza de Datos***

A continuación se eliminarán las variables nulas del dataframe o columnas que tienen un unico valor constante como la cantidad de objetos hurtados que siempre es uno dentro del archivo.

In [ ]:
med_hurto_df = med_hurto_df.drop(['cantidad','parentesco','nivel_academico','testigo','conducta','caracterizacion','modelo','grupo_bien'], axis=1)
med_hurto_df = med_hurto_df.drop(['articulo_penal','categoria_penal','permiso','unidad_medida','fecha_ingestion','color'], axis=1)
med_hurto_df.shape

(280870, 17)

Se cambia el tipo de dato de la columna "codigo_columna" a string ya que algunos datos los toma como númericos y otros cómo cadena de texto siendo los mismos 

In [ ]:
med_hurto_df['codigo_comuna'] = med_hurto_df['codigo_comuna'].astype(str)
comunas = med_hurto_df.codigo_comuna.unique()
comunas

array(['9', '4', '3', '13', '10', '11', '5', '6', '2', '16', '14', '8',
       '7', '12', '1', '15', 'SIN DATO', '90', '80', '70', '60', '50'],
      dtype=object)

In [ ]:
modalidades = med_hurto_df.modalidad.unique()
modalidades

array(['Descuido', 'Raponazo', 'Atraco', 'Cosquilleo',
       'Clonación de tarjeta', 'Engaño', 'Retención de dinero',
       'Rompimiento de ventana', 'Informático', 'Escopolamina',
       'Sin dato', 'Suplantación', 'Llamada millonaria',
       'Retención de tarjeta', 'Paquete chileno', 'Abuso de confianza',
       'Halado', 'Simulando necesidad', 'Rompimiento cerraduta',
       'Comisión de delito', 'Llave maestra', 'Fleteo', 'Clásica',
       'Tóxico o agente químico', 'Violencia intrafamiliar',
       'Enfrentamiento con la fuerza pública', 'Miedo o terror',
       'Vandalismo', 'Forcejeo', 'Auto robo'], dtype=object)

Se transformaron a minúsculas todas las columnas para evitar tener algunos barrios/localidades en mayúsculas y minúsculas y obtener datos errados en otra posible categoría.

In [ ]:
med_hurto_df['codigo_comuna'] = med_hurto_df['codigo_comuna'].str.lower()
med_hurto_df['codigo_comuna'] = med_hurto_df['codigo_comuna'].str.strip()
med_hurto_df['sexo'] = med_hurto_df['sexo'].str.lower()
med_hurto_df['sexo'] = med_hurto_df['sexo'].str.strip()
med_hurto_df['estado_civil'] = med_hurto_df['estado_civil'].str.lower()
med_hurto_df['estado_civil'] = med_hurto_df['estado_civil'].str.strip()
med_hurto_df['nombre_barrio'] = med_hurto_df['nombre_barrio'].str.lower()
med_hurto_df['nombre_barrio'] = med_hurto_df['nombre_barrio'].str.strip()
med_hurto_df['arma_medio'] = med_hurto_df['arma_medio'].str.lower()
med_hurto_df['arma_medio'] = med_hurto_df['arma_medio'].str.strip()
med_hurto_df['medio_transporte'] = med_hurto_df['medio_transporte'].str.lower()
med_hurto_df['medio_transporte'] = med_hurto_df['medio_transporte'].str.strip()
med_hurto_df['modalidad'] = med_hurto_df['modalidad'].str.lower()
med_hurto_df['modalidad'] = med_hurto_df['modalidad'].str.strip()
med_hurto_df['codigo_barrio'] = med_hurto_df['codigo_barrio'].str.lower()
med_hurto_df['codigo_barrio'] = med_hurto_df['codigo_barrio'].str.strip()
med_hurto_df['lugar'] = med_hurto_df['lugar'].str.lower()
med_hurto_df['lugar'] = med_hurto_df['lugar'].str.strip()
med_hurto_df['bien'] = med_hurto_df['bien'].str.lower()
med_hurto_df['bien'] = med_hurto_df['bien'].str.strip()
med_hurto_df['categoria_bien'] = med_hurto_df['categoria_bien'].str.lower()
med_hurto_df['categoria_bien'] = med_hurto_df['categoria_bien'].str.strip()
med_hurto_df['bien'] = med_hurto_df['bien'].replace('peso','efectivo')

# ****Preparación de los datos****

In [ ]:
# Se estadarizan las cadenas que tenemos como fechas para poder transformarlas correctamente
med_hurto_df['fecha_hecho'] = med_hurto_df['fecha_hecho'].apply(lambda x: x.replace('.000', ''))
med_hurto_df['fecha_hecho'] = med_hurto_df['fecha_hecho'].apply(lambda x: x.replace('T', ' '))
med_hurto_df['fecha_hecho'] = med_hurto_df['fecha_hecho'].apply(lambda x: x.replace('-05:00', ''))
med_hurto_df['fecha_hecho'] = pd.to_datetime(med_hurto_df['fecha_hecho'], format='%Y-%m-%d')

import calendar

med_hurto_df['dia'] = med_hurto_df['fecha_hecho'].apply(lambda x: x.day)
med_hurto_df['mes'] = med_hurto_df['fecha_hecho'].apply(lambda x: x.month)
med_hurto_df['anio'] = med_hurto_df['fecha_hecho'].apply(lambda x: x.year)
med_hurto_df['dia_semana'] = med_hurto_df['fecha_hecho'].dt.dayofweek

In [ ]:
# Se identifican los diferentes años para realizar el análsis
med_hurto_df['anio'].unique()

array([2012, 2013, 2014, 2015, 2016, 2020, 2019, 2018, 2021, 2022, 2017,
       2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011])

In [ ]:
# Se agrega un nuevo feature para saber si esa fecha correspondía a un día festivo
festivos_colombia_df = pd.read_csv('https://raw.githubusercontent.com/juanalvarez123/MINE-4101-proyecto-final/main/data/festivos_colombia_v2.csv')

festivos_colombia_df['fecha'] = pd.to_datetime(festivos_colombia_df['fecha'])
def is_holliday(fecha):
  is_holliday = festivos_colombia_df.loc[(festivos_colombia_df['fecha'] == fecha)]
  if (len(is_holliday) >= 1):
    return 1
  else:
    return 0

med_hurto_df['festivo'] = med_hurto_df.apply(lambda x: is_holliday(x.fecha_hecho), axis=1)

Para los días Festivos se agregó un listado de días festivos en Colombia y se validó día por día.

 0 : Día Festivo.

 1 : Dia NO festivo.


In [ ]:
# Se elimina el '#' del código del barrio
med_hurto_df['codigo_barrio'] = med_hurto_df['codigo_barrio'].apply(lambda x: x.replace('#', ''))

In [ ]:
# Se realiza clasificación de Hurtos
# Se define una lista de las modalidades que se considerarpon como peligrosas a pàrtir de esta se agrega una nueva columna en el dataframe.
modalidades_peligrosas = ['atraco', 'rompimiento de ventana', 'escopolamina', 'fleteo', 'tóxico o agente químico', 'miedo o terror', 'vandalismo', 'forcejeo', 'halado', 'rompimiento cerraduta']

def tipo_modalidad(modalidad):
  ndx = -1
  try:
    ndx = modalidades_peligrosas.index(modalidad)
  except:
    ndx = -1
  if ndx>=0:
    return 1 # Las modalidades peligrosas se identificarán con el número 1
  else:
    return 0 # Las modalidades NO peligrosas se identificarán con el número 0

med_hurto_df['tipo_modalidad'] = med_hurto_df.apply(lambda x: tipo_modalidad(x.modalidad), axis=1)

In [ ]:
# Clasificación Tipo de transporte
# Se define una lista con los tipos de transporte y estos son agregados a una nueva columna en el dataframe.
tipos_transporte = {'taxi': 0, 'autobus': 0, 'metro': 0, 'automóvil': 1, 'motocicleta': 1, 'motocicleta con parrillero': 1, 'bicicleta': 1, 'caminata': 2}

def tipo_transporte(tipo):
  ndx = -1
  try:
    return tipos_transporte[tipo]
  except:
    return 2

med_hurto_df['tipo_medio_transp'] = med_hurto_df.apply(lambda x: tipo_transporte(x.medio_transporte), axis=1)

Se definieron los siguientes grupos de transporte

Grupo 0 : Transporte público

Grupo 1 : Transporte particular.

Grupo 2 : Via pública

In [ ]:
# Clasificación Edades
def gruposEdad(edad):
  if (edad<12):
    return 0
  elif (edad> 11 and edad<19):
    return 1
  elif (edad>18 and edad<61):
    return 2
  else:
    return 3

med_hurto_df['grupo_edad'] = med_hurto_df.apply(lambda x: gruposEdad(x.edad), axis=1)

Se definieron los siguientes grupos de transporte

Grupo 0 : Niños 0 - 11 Años

Grupo 1 : Jovenes 12 - 18 Años

Grupo 2 : Adultos 19 - 60 Años

Grupo 3 : Adulto Mayor desde los 61 años

In [ ]:
# Clasificación Armas 
# Se define una lista de las armas que se consideran como peligrosas a partir de esta se agrega una nueva columna en el dataframe.
modalidades_peligrosas = ['arma de fuego', 'objeto contundente', 'arma cortopunzante', 'escopolamina', 'tóxico o agente químico']

def tipo_modalidad(modalidad):
  ndx = -1
  try:
    ndx = modalidades_peligrosas.index(modalidad)
  except:
    ndx = -1
  if ndx>=0:
    return 0 # Armas peligrosas se identificarán con el número 0
  else:
    return 1 # Armas NO peligrosas se identificarán con el número 1

med_hurto_df['tipo_arma'] = med_hurto_df.apply(lambda x: tipo_modalidad(x.modalidad), axis=1)

Se definieron los siguientes grupos de armas:

Grupo 0 : Armas peligrosas

Grupo 1 : Armas NO peligrosas

In [ ]:
# Aquí se deben hacer los agrupamientos y el count (cnt)
med_hurto_group_df = med_hurto_df.groupby(['codigo_comuna','codigo_barrio','tipo_modalidad','dia_semana','festivo','tipo_medio_transp','tipo_arma','sexo','grupo_edad']).agg({'fecha_hecho': ['count']})
med_hurto_group_df.reset_index(inplace=True)
med_hurto_group_df.columns = ['codigo_comuna','codigo_barrio','tipo_modalidad','dia_semana','festivo','tipo_medio_transp','tipo_arma','sexo','grupo_edad','hurtos']
med_hurto_group_df

codigo_comuna codigo_barrio  tipo_modalidad  dia_semana  festivo  \
0                 1          0101               0           0        0   
1                 1          0101               0           0        0   
2                 1          0101               0           0        0   
3                 1          0101               0           0        0   
4                 1          0101               0           0        0   
...             ...           ...             ...         ...      ...   
26491      sin dato      sin dato               1           6        0   
26492      sin dato      sin dato               1           6        0   
26493      sin dato      sin dato               1           6        0   
26494      sin dato      sin dato               1           6        0   
26495      sin dato      sin dato               1           6        0   

       tipo_medio_transp  tipo_arma    sexo  grupo_edad  hurtos  
0                      0          1  hombre           2       2  
1                      0          1   mujer           2       3  
2                      2          1  hombre           1       1  
3                      2          1  hombre           2       8  
4                      2          1  hombre           3       3  
...                  ...        ...     ...         ...     ...  
26491                  0          1   mujer           2       3  
26492                  1          1  hombre           2       3  
26493                  1          1   mujer           2       3  
26494                  2          1  hombre           2       4  
26495                  2          1   mujer           2       4  

[26496 rows x 10 columns]

# ****Selección de Modelos****

## Regresión Lineal

In [ ]:
med_hurto_group_df = med_hurto_group_df.drop('sexo', axis=1)

temp_med_hurto_df = med_hurto_group_df[(med_hurto_group_df['codigo_comuna'] != 'sin dato')]
new_med_hurto_df = temp_med_hurto_df[(temp_med_hurto_df['codigo_barrio'] != 'sin dato 90 santa elena')
& (temp_med_hurto_df['codigo_barrio'] != 'sin dato 60 san cristobal')
& (temp_med_hurto_df['codigo_barrio'] != 'sin dato 80 san antonio de prado')
& (temp_med_hurto_df['codigo_barrio'] != 'sin dato 10 la candelaria')]

#new_med_hurto_df = new_med_hurto_df.drop('codigo_barrio', axis=1)

Variables para entrenar el modelo

In [ ]:
features = ['tipo_modalidad','dia_semana','festivo','tipo_medio_transp','tipo_arma','grupo_edad']

In [ ]:
new_med_hurto_df[features + ['hurtos']].corr()

tipo_modalidad  dia_semana   festivo  tipo_medio_transp  \
tipo_modalidad           1.000000    0.017461 -0.020650           0.009157   
dia_semana               0.017461    1.000000 -0.039773           0.018154   
festivo                 -0.020650   -0.039773  1.000000           0.045163   
tipo_medio_transp        0.009157    0.018154  0.045163           1.000000   
tipo_arma               -0.174390   -0.043000  0.004412          -0.050170   
grupo_edad              -0.041404   -0.007746  0.001016          -0.103488   
hurtos                   0.020284   -0.003760 -0.020378           0.105939   

                   tipo_arma  grupo_edad    hurtos  
tipo_modalidad     -0.174390   -0.041404  0.020284  
dia_semana         -0.043000   -0.007746 -0.003760  
festivo             0.004412    0.001016 -0.020378  
tipo_medio_transp  -0.050170   -0.103488  0.105939  
tipo_arma           1.000000   -0.040993  0.031599  
grupo_edad         -0.040993    1.000000  0.035129  
hurtos              0.031599    0.035129  1.000000

Creación de Datasets de entrenamiento y pruebas

In [ ]:
scaler = StandardScaler()

In [ ]:
X = new_med_hurto_df[features]

In [ ]:
Y = new_med_hurto_df['hurtos']

In [ ]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 1)

In [ ]:
X_train.shape

(21104, 6)

In [ ]:
X_test.shape

(5276, 6)

In [ ]:
X_train_sc = scaler.fit_transform(X_train)
X_test_sc = scaler.transform(X_test)

Entrenamiento del modelo

In [ ]:
regr = LinearRegression()

In [ ]:
regr.fit(X_train_sc, Y_train)

LinearRegression()

In [ ]:
regr.coef_

array([ 1.11759878, -0.10069865, -0.8970104 ,  4.07372971,  1.68113097,
        1.78294822])

In [ ]:
regr.intercept_

10.642721758908264

Evaluando el modelo

In [ ]:
preds_train = regr.predict(X_train_sc)
preds_test = regr.predict(X_test_sc)

In [ ]:
mean_absolute_error(Y_train, preds_train), mean_absolute_error(Y_test, preds_test)

(11.35428069247011, 11.237129089884425)

# ****Conclusión del Modelo****

De acuerdo a alas caracteristicas del dataset, la variable objetivo y el resultado obtenido en error en entrenamiento y test de 11.35 y 11.23, concluimos que teniamos un resultado con underfitting y que era muy complejo predecir el número de hurtos en algunos barrios ya que los datos en estos eran pocos por ejemplo algunos barrios para un mes tenían 5 registros con hurtos entre 1 - 5, comparados con otros barrios que alcanzaban a llegar a 38 hurtos, verificando tambien el análisis exploratorio realizado en la primera entrega: 

<img src="https://lh3.google.com/u/0/d/1-caUvCXaPrnt5DYd1MmBQIt9u4wIOMAv=w1920-h859-iv1" ><br> 

Los datos de las diferentes Zonas y en especial de los barrios no eran lo suficientemente representativos para realizar el análisis. 

### Bibliografía

<sup>1</sup> Hurto a persona - Alcaldía de Medellín http://medata.gov.co/dataset/hurto-persona